lib/pos_mempool.go

```diff
@@ -0,0 +1,418 @@
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+	"math"
+	"math/big"
+)
+
+// ========================
+//	TransactionRegister
+// ========================
+
+// TransactionRegister is the primary transaction store of the PoS Mempool. The register is responsible for determining
+// the Fee-Time ordering of transactions. The operations supported by the register are: adding a transaction, removing
+// a transaction, iterating through all transactions in fee-time order, and retrieving all transactions ordered in
+// fee-time. The TransactionRegister doesn't perform any validation on the transactions, it just accepts the provided
+// MempoolTx and adds it to the appropriate FeeTimeBucket.
+type TransactionRegister struct {
+	// feeTimeBucketSet is a set of FeeTimeBucket objects. The set is ordered by the FeeTimeBucket's ranges, based on FeeTimeBucketComparator.
+	feeTimeBucketSet *treeset.Set
+	// feeTimeBucketsByMinFeeMap is a map of FeeTimeBucket minimum fees to FeeTimeBucket objects. It is used to quickly find
+	// a FeeTimeBucket given its fee. Note that using a map here is preferable over e.g. a slice, since the index of a
+	// FeeTimeBucket is not necessarily its position in the set.
+	feeTimeBucketsByMinFeeMap map[uint64]*FeeTimeBucket
+	// txnMembership is a set of transaction hashes. It is used to determine existence of a transaction in the register.
+	txnMembership *Set[BlockHash]
+	// totalTxnSize is the total size of all transactions in the register.
+	totalTxnSize uint64
+
+	params *DeSoParams
+
+	// minimumNetworkFeeNanosPerKB is the base fee rate for the lowest fee FeeTimeBucket. This value corresponds to
+	// GlobalParamsEntry's MinimumNetworkFeeNanosPerKB.
+	minimumNetworkFeeNanosPerKB *big.Float
+	// feeBucketRateMultiplierBasisPoints is the fee rate multiplier for FeeTimeBucket objects. This value corresponds to
+	// GlobalParamsEntry's FeeBucketRateMultiplierBasisPoints.
+	feeBucketRateMultiplierBasisPoints *big.Float
+}
+
+func NewTransactionRegister(params *DeSoParams, globalParams *GlobalParamsEntry) *TransactionRegister {
+	feeTimeBucketSet := treeset.NewWith(FeeTimeBucketComparator)
+	minNetworkFee, bucketMultiplier := globalParams.ComputeFeeTimeBucketMinimumFeeAndMultiplier()
+
+	return &TransactionRegister{
+		feeTimeBucketSet:                   feeTimeBucketSet,
+		feeTimeBucketsByMinFeeMap:          make(map[uint64]*FeeTimeBucket),
+		txnMembership:                      NewSet([]BlockHash{}),
+		totalTxnSize:                       0,
+		params:                             params,
+		minimumNetworkFeeNanosPerKB:        minNetworkFee,
+		feeBucketRateMultiplierBasisPoints: bucketMultiplier,
+	}
+}
+
+// FeeTimeBucketComparator is a comparator function for FeeTimeBucket objects. It is used to order FeeTimeBucket objects
+// in the TransactionRegister's feeTimeBucketSet based on fee ranges (higher fee ranges are ordered first).
+func FeeTimeBucketComparator(a, b interface{}) int {
```

Comment:
Nit, scoping. This function is an implementation detail of the TransactionRegister, so we don't want to be expored outside of the current package.
```suggestion
func feeTimeBucketComparator(a, b interface{}) int {
```